core/src/test/scala/integration/kafka/api/ClearPurgatoryOnLeaderMovementTest.scala

```diff
@@ -0,0 +1,88 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package integration.kafka.api
+
+import java.util.Properties
+import java.util.concurrent.ExecutionException
+
+import kafka.admin.{PreferredReplicaLeaderElectionCommand, ReassignPartitionsCommand}
+import kafka.common.TopicAndPartition
+import kafka.integration.KafkaServerTestHarness
+import kafka.server.KafkaConfig
+import kafka.utils.TestUtils
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.errors.NotLeaderForPartitionException
+import org.junit.Assert.{assertFalse, assertTrue}
+import org.junit.Test
+
+/**
+ * Test the case where produce requests in the purgatory should be cleared when the leader migrates to another broker.
+ */
+class ClearPurgatoryOnLeaderMovementTest extends KafkaServerTestHarness {
```

Comment:
I haven't checked carefully enough if this is possible, but could you just mock out the `ReplicaManager` and avoid the full test harness? and if so, can you fold this into the `ReplicaManagerTest`